coderd/authzquery/authz.go

```diff
@@ -0,0 +1,314 @@
+package authzquery
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"cdr.dev/slog"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/coderd/rbac"
+)
+
+// TODO:
+// - We need to handle authorizing the CRUD of objects with RBAC being related
+//   to some other object. Eg: workspace builds, group members, etc.
+
+var (
+	// NoActorError wraps ErrNoRows for the api to return a 404. This is the correct
+	// response when the user is not authorized.
+	NoActorError = xerrors.Errorf("no authorization actor in context: %w", sql.ErrNoRows)
+)
+
+// NotAuthorizedError is a sentinel error that unwraps to sql.ErrNoRows.
+// This allows the internal error to be read by the caller if needed. Otherwise
+// it will be handled as a 404.
+type NotAuthorizedError struct {
+	Err error
+}
+
+func (e NotAuthorizedError) Error() string {
+	return fmt.Sprintf("unauthorized: %s", e.Err.Error())
+}
+
+// Unwrap will always unwrap to a sql.ErrNoRows so the API returns a 404.
+// So 'errors.Is(err, sql.ErrNoRows)' will always be true.
+func (e NotAuthorizedError) Unwrap() error {
+	return sql.ErrNoRows
+}
+
+func LogNotAuthorizedError(ctx context.Context, logger slog.Logger, err error) error {
+	// Only log the errors if it is an UnauthorizedError error.
+	internalError := new(rbac.UnauthorizedError)
+	if err != nil && xerrors.As(err, internalError) {
+		logger.Debug(ctx, "unauthorized",
+			slog.F("internal", internalError.Internal()),
+			slog.F("input", internalError.Input()),
+			slog.Error(err),
+		)
+	}
+	return NotAuthorizedError{
+		Err: err,
+	}
+}
+
+func authorizedInsert[ArgumentType any,
+	Insert func(ctx context.Context, arg ArgumentType) error](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
```

Comment:
I wonder if it is really a good idea to parameterize the action here.  In all sensible cases, the action _should_ be `ActionCreate` and if it is not, it is a code smell and maybe we shouldn't be creating shortcuts for that sort of thing?

Reply:
You are right, there is like 1 or two edge cases. We can probably handle these manually.

Reply:
Fixed, no aciton is passed. `ActionCreate` is used.